File: services/users-service/internal/repository/user_repository.go
package repository

import (
	"context"
	"encoding/json"
	"fmt"
	"math"
	"strings"
	"time"

	"github.com/google/uuid"
	"github.com/jackc/pgx/v5"
	"github.com/jackc/pgx/v5/pgxpool"

	"github.com/AnathanWang/andexevents/services/users-service/internal/model"
)

// UserRepository интерфейс репозитория пользователей
type UserRepository interface {
	Create(ctx context.Context, supabaseUID, email string, displayName, photoURL *string) (*model.User, error)
	GetByID(ctx context.Context, id string) (*model.User, error)
	GetBySupabaseUID(ctx context.Context, supabaseUID string) (*model.User, error)
	GetByEmail(ctx context.Context, email string) (*model.User, error)
	Update(ctx context.Context, id string, req *model.UpdateUserRequest) (*model.User, error)
	UpdateLocation(ctx context.Context, id string, latitude, longitude float64) error
	GetMatches(ctx context.Context, userID string, latitude, longitude *float64, radiusKm float64, limit int) ([]model.User, error)
	UpdateSupabaseUID(ctx context.Context, id, supabaseUID string) (*model.User, error)
}

type userRepository struct {
	pool *pgxpool.Pool
}

// NewUserRepository создаёт новый репозиторий пользователей
func NewUserRepository(pool *pgxpool.Pool) UserRepository {
	return &userRepository{pool: pool}
}

// Create создаёт нового пользователя
func (r *userRepository) Create(ctx context.Context, supabaseUID, email string, displayName, photoURL *string) (*model.User, error) {
	id := uuid.New().String()
	now := time.Now()

	query := `
		INSERT INTO "User" (
			id, "supabaseUid", email, "displayName", "photoUrl",
			role, "isProfileVisible", "isLocationVisible", "maxDistance",
			"isOnboardingCompleted", "createdAt", "updatedAt"
		) VALUES (
			$1, $2, $3, $4, $5,
			'USER', true, true, 50000,
			false, $6, $6
		)
		RETURNING id, "supabaseUid", email, "displayName", "photoUrl", bio, interests,
			"socialLinks", age, gender, role, "lastLatitude", "lastLongitude",
			"lastLocationUpdate", "isProfileVisible", "isLocationVisible",
			"minAge", "maxAge", "maxDistance", "fcmToken", "isOnboardingCompleted",
			"createdAt", "updatedAt"
	`

	var user model.User
	var interests []string
	var socialLinks []byte

	err := r.pool.QueryRow(ctx, query, id, supabaseUID, email, displayName, photoURL, now).Scan(
		&user.ID, &user.SupabaseUID, &user.Email, &user.DisplayName, &user.PhotoURL,
		&user.Bio, &interests, &socialLinks, &user.Age, &user.Gender, &user.Role,
		&user.LastLatitude, &user.LastLongitude, &user.LastLocationUpdate,
		&user.IsProfileVisible, &user.IsLocationVisible, &user.MinAge, &user.MaxAge,
		&user.MaxDistance, &user.FCMToken, &user.IsOnboardingCompleted,
		&user.CreatedAt, &user.UpdatedAt,
	)
	if err != nil {
		return nil, err
	}

	user.Interests = interests
	if socialLinks != nil {
		user.SocialLinks = socialLinks
	}

	return &user, nil
}

// GetByID получает пользователя по ID
func (r *userRepository) GetByID(ctx context.Context, id string) (*model.User, error) {
	query := `
		SELECT id, "supabaseUid", email, "displayName", "photoUrl", bio, interests,
			"socialLinks", age, gender, role, "lastLatitude", "lastLongitude",
			"lastLocationUpdate", "isProfileVisible", "isLocationVisible",
			"minAge", "maxAge", "maxDistance", "fcmToken", "isOnboardingCompleted",
			"createdAt", "updatedAt"
		FROM "User"
		WHERE id = $1
	`

	var user model.User
	var interests []string
	var socialLinks []byte

	err := r.pool.QueryRow(ctx, query, id).Scan(
		&user.ID, &user.SupabaseUID, &user.Email, &user.DisplayName, &user.PhotoURL,
		&user.Bio, &interests, &socialLinks, &user.Age, &user.Gender, &user.Role,
		&user.LastLatitude, &user.LastLongitude, &user.LastLocationUpdate,
		&user.IsProfileVisible, &user.IsLocationVisible, &user.MinAge, &user.MaxAge,
		&user.MaxDistance, &user.FCMToken, &user.IsOnboardingCompleted,
		&user.CreatedAt, &user.UpdatedAt,
	)

	if err == pgx.ErrNoRows {
		return nil, nil
	}
	if err != nil {
		return nil, err
	}

	user.Interests = interests
	if socialLinks != nil {
		user.SocialLinks = socialLinks
	}

	return &user, nil
}

// GetBySupabaseUID получает пользователя по Supabase UID
func (r *userRepository) GetBySupabaseUID(ctx context.Context, supabaseUID string) (*model.User, error) {
	query := `
		SELECT id, "supabaseUid", email, "displayName", "photoUrl", bio, interests,
			"socialLinks", age, gender, role, "lastLatitude", "lastLongitude",
			"lastLocationUpdate", "isProfileVisible", "isLocationVisible",
			"minAge", "maxAge", "maxDistance", "fcmToken", "isOnboardingCompleted",
			"createdAt", "updatedAt"
		FROM "User"
		WHERE "supabaseUid" = $1
	`

	var user model.User
	var interests []string
	var socialLinks []byte

	err := r.pool.QueryRow(ctx, query, supabaseUID).Scan(
		&user.ID, &user.SupabaseUID, &user.Email, &user.DisplayName, &user.PhotoURL,
		&user.Bio, &interests, &socialLinks, &user.Age, &user.Gender, &user.Role,
		&user.LastLatitude, &user.LastLongitude, &user.LastLocationUpdate,
		&user.IsProfileVisible, &user.IsLocationVisible, &user.MinAge, &user.MaxAge,
		&user.MaxDistance, &user.FCMToken, &user.IsOnboardingCompleted,
		&user.CreatedAt, &user.UpdatedAt,
	)

	if err == pgx.ErrNoRows {
		return nil, nil
	}
	if err != nil {
		return nil, err
	}

	user.Interests = interests
	if socialLinks != nil {
		user.SocialLinks = socialLinks
	}

	return &user, nil
}

// GetByEmail получает пользователя по email
func (r *userRepository) GetByEmail(ctx context.Context, email string) (*model.User, error) {
	query := `
		SELECT id, "supabaseUid", email, "displayName", "photoUrl", bio, interests,
			"socialLinks", age, gender, role, "lastLatitude", "lastLongitude",
			"lastLocationUpdate", "isProfileVisible", "isLocationVisible",
			"minAge", "maxAge", "maxDistance", "fcmToken", "isOnboardingCompleted",
			"createdAt", "updatedAt"
		FROM "User"
		WHERE email = $1
	`

	var user model.User
	var interests []string
	var socialLinks []byte

	err := r.pool.QueryRow(ctx, query, email).Scan(
		&user.ID, &user.SupabaseUID, &user.Email, &user.DisplayName, &user.PhotoURL,
		&user.Bio, &interests, &socialLinks, &user.Age, &user.Gender, &user.Role,
		&user.LastLatitude, &user.LastLongitude, &user.LastLocationUpdate,
		&user.IsProfileVisible, &user.IsLocationVisible, &user.MinAge, &user.MaxAge,
		&user.MaxDistance, &user.FCMToken, &user.IsOnboardingCompleted,
		&user.CreatedAt, &user.UpdatedAt,
	)

	if err == pgx.ErrNoRows {
		return nil, nil
	}
	if err != nil {
		return nil, err
	}

	user.Interests = interests
	if socialLinks != nil {
		user.SocialLinks = socialLinks
	}

	return &user, nil
}

// UpdateSupabaseUID обновляет supabaseUid у существующего пользователя
func (r *userRepository) UpdateSupabaseUID(ctx context.Context, id, supabaseUID string) (*model.User, error) {
	query := `
		UPDATE "User"
		SET "supabaseUid" = $1, "updatedAt" = $2
		WHERE id = $3
		RETURNING id, "supabaseUid", email, "displayName", "photoUrl", bio, interests,
			"socialLinks", age, gender, role, "lastLatitude", "lastLongitude",
			"lastLocationUpdate", "isProfileVisible", "isLocationVisible",
			"minAge", "maxAge", "maxDistance", "fcmToken", "isOnboardingCompleted",
			"createdAt", "updatedAt"
	`

	var user model.User
	var interests []string
	var socialLinks []byte

	err := r.pool.QueryRow(ctx, query, supabaseUID, time.Now(), id).Scan(
		&user.ID, &user.SupabaseUID, &user.Email, &user.DisplayName, &user.PhotoURL,
		&user.Bio, &interests, &socialLinks, &user.Age, &user.Gender, &user.Role,
		&user.LastLatitude, &user.LastLongitude, &user.LastLocationUpdate,
		&user.IsProfileVisible, &user.IsLocationVisible, &user.MinAge, &user.MaxAge,
		&user.MaxDistance, &user.FCMToken, &user.IsOnboardingCompleted,
		&user.CreatedAt, &user.UpdatedAt,
	)
	if err != nil {
		return nil, err
	}

	user.Interests = interests
	if socialLinks != nil {
		user.SocialLinks = socialLinks
	}

	return &user, nil
}

// Update обновляет профиль пользователя
func (r *userRepository) Update(ctx context.Context, id string, req *model.UpdateUserRequest) (*model.User, error) {
	// Собираем динамический UPDATE запрос
	setClauses := []string{`"updatedAt" = $1`}
	args := []interface{}{time.Now()}
	argNum := 2

	if req.DisplayName != nil {
		setClauses = append(setClauses, fmt.Sprintf(`"displayName" = $%d`, argNum))
		args = append(args, *req.DisplayName)
		argNum++
	}

	if req.PhotoURL != nil {
		setClauses = append(setClauses, fmt.Sprintf(`"photoUrl" = $%d`, argNum))
		args = append(args, *req.PhotoURL)
		argNum++
	}

	if req.Bio != nil {
		setClauses = append(setClauses, fmt.Sprintf(`bio = $%d`, argNum))
		args = append(args, *req.Bio)
		argNum++
	}

	if req.Age != nil {
		setClauses = append(setClauses, fmt.Sprintf(`age = $%d`, argNum))
		args = append(args, *req.Age)
		argNum++
	}

	if req.Gender != nil {
		setClauses = append(setClauses, fmt.Sprintf(`gender = $%d`, argNum))
		args = append(args, *req.Gender)
		argNum++
	}

	if req.Interests != nil {
		setClauses = append(setClauses, fmt.Sprintf(`interests = $%d`, argNum))
		args = append(args, req.Interests)
		argNum++
	}

	if req.SocialLinks != nil {
		setClauses = append(setClauses, fmt.Sprintf(`"socialLinks" = $%d`, argNum))
		args = append(args, req.SocialLinks)
		argNum++
	}

	if req.IsOnboardingCompleted != nil {
		setClauses = append(setClauses, fmt.Sprintf(`"isOnboardingCompleted" = $%d`, argNum))
		args = append(args, *req.IsOnboardingCompleted)
		argNum++
	}

	// Добавляем id в конец
	args = append(args, id)

	query := fmt.Sprintf(`
		UPDATE "User"
		SET %s
		WHERE id = $%d
		RETURNING id, "supabaseUid", email, "displayName", "photoUrl", bio, interests,
			"socialLinks", age, gender, role, "lastLatitude", "lastLongitude",
			"lastLocationUpdate", "isProfileVisible", "isLocationVisible",
			"minAge", "maxAge", "maxDistance", "fcmToken", "isOnboardingCompleted",
			"createdAt", "updatedAt"
	`, strings.Join(setClauses, ", "), argNum)

	var user model.User
	var interests []string
	var socialLinks []byte

	err := r.pool.QueryRow(ctx, query, args...).Scan(
		&user.ID, &user.SupabaseUID, &user.Email, &user.DisplayName, &user.PhotoURL,
		&user.Bio, &interests, &socialLinks, &user.Age, &user.Gender, &user.Role,
		&user.LastLatitude, &user.LastLongitude, &user.LastLocationUpdate,
		&user.IsProfileVisible, &user.IsLocationVisible, &user.MinAge, &user.MaxAge,
		&user.MaxDistance, &user.FCMToken, &user.IsOnboardingCompleted,
		&user.CreatedAt, &user.UpdatedAt,
	)
	if err != nil {
		return nil, err
	}

	user.Interests = interests
	if socialLinks != nil {
		user.SocialLinks = socialLinks
	}

	return &user, nil
}

// UpdateLocation обновляет геолокацию пользователя
func (r *userRepository) UpdateLocation(ctx context.Context, id string, latitude, longitude float64) error {
	query := `
		UPDATE "User"
		SET "lastLatitude" = $1, "lastLongitude" = $2, "lastLocationUpdate" = $3, "updatedAt" = $3
		WHERE id = $4
	`

	now := time.Now()
	_, err := r.pool.Exec(ctx, query, latitude, longitude, now, id)
	return err
}

// GetMatches получает список потенциальных матчей
func (r *userRepository) GetMatches(ctx context.Context, userID string, latitude, longitude *float64, radiusKm float64, limit int) ([]model.User, error) {
	// Получаем текущего пользователя
	currentUser, err := r.GetByID(ctx, userID)
	if err != nil {
		return nil, err
	}
	if currentUser == nil {
		return []model.User{}, nil
	}

	// Используем переданные координаты или последние известные
	userLat := latitude
	userLon := longitude
	if userLat == nil {
		userLat = currentUser.LastLatitude
	}
	if userLon == nil {
		userLon = currentUser.LastLongitude
	}

	// Если координаты недоступны, возвращаем пустой список
	if userLat == nil || userLon == nil {
		return []model.User{}, nil
	}

	// Вычисляем границы поиска
	radiusMeters := radiusKm * 1000
	earthRadiusKm := 6371.0
	latChange := (radiusKm / earthRadiusKm) * (180.0 / math.Pi)
	lonChange := (radiusKm / (earthRadiusKm * math.Cos(*userLat*math.Pi/180.0))) * (180.0 / math.Pi)

	minLat := *userLat - latChange
	maxLat := *userLat + latChange
	minLon := *userLon - lonChange
	maxLon := *userLon + lonChange

	query := `
		SELECT id, "supabaseUid", email, "displayName", "photoUrl", bio, interests,
			"socialLinks", age, gender, role, "lastLatitude", "lastLongitude",
			"lastLocationUpdate", "isProfileVisible", "isLocationVisible",
			"minAge", "maxAge", "maxDistance", "fcmToken", "isOnboardingCompleted",
			"createdAt", "updatedAt"
		FROM "User"
		WHERE id != $1
			AND "isOnboardingCompleted" = true
			AND "isProfileVisible" = true
			AND "lastLatitude" IS NOT NULL
			AND "lastLongitude" IS NOT NULL
			AND "lastLatitude" >= $2 AND "lastLatitude" <= $3
			AND "lastLongitude" >= $4 AND "lastLongitude" <= $5
	`

	args := []interface{}{userID, minLat, maxLat, minLon, maxLon}
	argNum := 6

	// Добавляем возрастной фильтр если установлен
	if currentUser.MinAge != nil {
		query += fmt.Sprintf(` AND age >= $%d`, argNum)
		args = append(args, *currentUser.MinAge)
		argNum++
	}
	if currentUser.MaxAge != nil {
		query += fmt.Sprintf(` AND age <= $%d`, argNum)
		args = append(args, *currentUser.MaxAge)
		argNum++
	}

	query += fmt.Sprintf(` LIMIT $%d`, argNum)
	args = append(args, limit)

	rows, err := r.pool.Query(ctx, query, args...)
	if err != nil {
		return nil, err
	}
	defer rows.Close()

	var users []model.User
	for rows.Next() {
		var user model.User
		var interests []string
		var socialLinks []byte

		err := rows.Scan(
			&user.ID, &user.SupabaseUID, &user.Email, &user.DisplayName, &user.PhotoURL,
			&user.Bio, &interests, &socialLinks, &user.Age, &user.Gender, &user.Role,
			&user.LastLatitude, &user.LastLongitude, &user.LastLocationUpdate,
			&user.IsProfileVisible, &user.IsLocationVisible, &user.MinAge, &user.MaxAge,
			&user.MaxDistance, &user.FCMToken, &user.IsOnboardingCompleted,
			&user.CreatedAt, &user.UpdatedAt,
		)
		if err != nil {
			return nil, err
		}

		user.Interests = interests
		if socialLinks != nil {
			user.SocialLinks = socialLinks
		}

		// Вычисляем расстояние
		if user.LastLatitude != nil && user.LastLongitude != nil {
			dist := haversineDistance(*userLat, *userLon, *user.LastLatitude, *user.LastLongitude)
			// Фильтруем по точному расстоянию (более точно чем bounding box)
			if dist <= radiusMeters {
				user.Distance = &dist
				users = append(users, user)
			}
		}
	}

	return users, nil
}

// haversineDistance вычисляет расстояние между двумя точками в метрах
func haversineDistance(lat1, lon1, lat2, lon2 float64) float64 {
	const earthRadiusMeters = 6371000.0

	lat1Rad := lat1 * math.Pi / 180
	lat2Rad := lat2 * math.Pi / 180
	deltaLat := (lat2 - lat1) * math.Pi / 180
	deltaLon := (lon2 - lon1) * math.Pi / 180

	a := math.Sin(deltaLat/2)*math.Sin(deltaLat/2) +
		math.Cos(lat1Rad)*math.Cos(lat2Rad)*math.Sin(deltaLon/2)*math.Sin(deltaLon/2)
	c := 2 * math.Atan2(math.Sqrt(a), math.Sqrt(1-a))

	return earthRadiusMeters * c
}

// ScanUser helper для сканирования пользователя
func ScanUser(row pgx.Row) (*model.User, error) {
	var user model.User
	var interests []string
	var socialLinks []byte

	err := row.Scan(
		&user.ID, &user.SupabaseUID, &user.Email, &user.DisplayName, &user.PhotoURL,
		&user.Bio, &interests, &socialLinks, &user.Age, &user.Gender, &user.Role,
		&user.LastLatitude, &user.LastLongitude, &user.LastLocationUpdate,
		&user.IsProfileVisible, &user.IsLocationVisible, &user.MinAge, &user.MaxAge,
		&user.MaxDistance, &user.FCMToken, &user.IsOnboardingCompleted,
		&user.CreatedAt, &user.UpdatedAt,
	)

	if err == pgx.ErrNoRows {
		return nil, nil
	}
	if err != nil {
		return nil, err
	}

	user.Interests = interests
	if socialLinks != nil {
		user.SocialLinks = json.RawMessage(socialLinks)
	}

	return &user, nil
}

File: services-java/users-service/src/main/java/com/andexevents/users/repo/UserRepository.java
package com.andexevents.users.repo;

import com.andexevents.users.model.GlobalMatchContext;
import com.andexevents.users.model.UserDto;
import com.andexevents.users.service.UserService;
import com.fasterxml.jackson.core.type.TypeReference;
import com.fasterxml.jackson.databind.ObjectMapper;
import org.springframework.jdbc.core.JdbcTemplate;
import org.springframework.jdbc.core.RowMapper;
import org.springframework.stereotype.Repository;

import java.sql.Array;
import java.sql.ResultSet;
import java.sql.SQLException;
import java.time.Instant;
import java.util.*;
import java.util.UUID;

@Repository
public class UserRepository {
    private final JdbcTemplate jdbcTemplate;
    private final ObjectMapper objectMapper;

    public UserRepository(JdbcTemplate jdbcTemplate, ObjectMapper objectMapper) {
        this.jdbcTemplate = jdbcTemplate;
        this.objectMapper = objectMapper;
    }

    public Optional<UserDto> findByEmail(String email) {
        List<UserDto> rows = jdbcTemplate.query(
                "SELECT * FROM users.\"User\" WHERE email = ?",
                mapper(),
                email
        );
        return rows.stream().findFirst();
    }

    public Optional<UserDto> findById(String id) {
        List<UserDto> rows = jdbcTemplate.query(
                "SELECT * FROM users.\"User\" WHERE id = ?",
                mapper(),
                id
        );
        return rows.stream().findFirst();
    }

    public List<UserDto> findAllForModeration() {
        return jdbcTemplate.query(
                "SELECT * FROM users.\"User\" ORDER BY \"createdAt\" DESC",
                mapper()
        );
    }

    public Optional<UserDto> findByFirebaseUid(String firebaseUid) {
        List<UserDto> rows = jdbcTemplate.query(
                "SELECT * FROM users.\"User\" WHERE \"firebaseUid\" = ? OR \"supabaseUid\" = ?",
                mapper(),
                firebaseUid,
                firebaseUid
        );
        return rows.stream().findFirst();
    }

    public UserDto insertUser(String id, String firebaseUid, String email, String displayName, String photoUrl) {
        jdbcTemplate.update(
                "INSERT INTO users.\"User\" (id, \"firebaseUid\", \"supabaseUid\", email, \"displayName\", \"photoUrl\", \"createdAt\", \"updatedAt\") VALUES (?, ?, ?, ?, ?, ?, NOW(), NOW())",
                id,
                firebaseUid,
                firebaseUid,
                email,
                displayName,
                photoUrl
        );
        return findById(id).orElseThrow();
    }

    public UserDto updateFirebaseUid(String userId, String firebaseUid) {
        jdbcTemplate.update(
                "UPDATE users.\"User\" SET \"firebaseUid\" = ?, \"supabaseUid\" = COALESCE(\"supabaseUid\", ?), \"updatedAt\" = NOW() WHERE id = ?",
                firebaseUid,
                firebaseUid,
                userId
        );
        return findById(userId).orElseThrow();
    }

    public UserDto updateProfile(String userId, Map<String, Object> updates) {
        if (updates.isEmpty()) {
            return findById(userId).orElseThrow();
        }

        List<Object> params = new ArrayList<>();
        StringBuilder sql = new StringBuilder("UPDATE users.\"User\" SET ");

        int i = 0;
        for (Map.Entry<String, Object> e : updates.entrySet()) {
            if (i++ > 0) sql.append(", ");
            sql.append('"').append(e.getKey()).append('"').append(" = ");

            if ("socialLinks".equals(e.getKey())) {
                sql.append("CAST(? AS jsonb)");
                try {
                    params.add(objectMapper.writeValueAsString(e.getValue()));
                } catch (Exception ex) {
                    throw new IllegalArgumentException("Failed to serialize socialLinks", ex);
                }
            } else {
                sql.append("?");
                params.add(e.getValue());
            }
        }
        sql.append(", \"updatedAt\" = NOW() WHERE id = ?");
        params.add(userId);

        jdbcTemplate.update(sql.toString(), params.toArray());
        return findById(userId).orElseThrow();
    }

    public Map<String, Integer> findSharedGoingEventCounts(String userId, List<String> candidateIds) {
        if (candidateIds == null || candidateIds.isEmpty()) {
            return Map.of();
        }

        String placeholders = String.join(", ", Collections.nCopies(candidateIds.size(), "?"));
        String sql = """
                SELECT p2."userId" AS candidate_id, COUNT(DISTINCT p2."eventId") AS shared_count
                FROM events."Participant" p1
                JOIN events."Participant" p2 ON p1."eventId" = p2."eventId"
                WHERE p1."userId" = ?
                  AND p2."userId" <> p1."userId"
                  AND p1.status = 'GOING'::events."ParticipantStatus"
                  AND p2.status = 'GOING'::events."ParticipantStatus"
                  AND p2."userId" IN (%s)
                GROUP BY p2."userId"
                """.formatted(placeholders);

        List<Object> params = new ArrayList<>();
        params.add(userId);
        params.addAll(candidateIds);

        Map<String, Integer> counts = new HashMap<>();
        try {
            jdbcTemplate.query(
                    sql,
                    rs -> {
                        counts.put(rs.getString("candidate_id"), rs.getInt("shared_count"));
                    },
                    params.toArray()
            );
        } catch (Exception ignored) {
            // events schema may be unavailable in isolated test DB
            return Map.of();
        }
        return counts;
    }

    public void updateLocation(String userId, double latitude, double longitude) {
        jdbcTemplate.update(
                "UPDATE users.\"User\" SET \"lastLatitude\" = ?, \"lastLongitude\" = ?, \"lastLocationUpdate\" = NOW(), \"updatedAt\" = NOW() WHERE id = ?",
                latitude,
                longitude,
                userId
        );
    }

    public GlobalMatchContext findGlobalMatchContext(String userId) {
        String sql = """
                SELECT
                  CASE WHEN "userAId" = ? THEN "userBId" ELSE "userAId" END AS other_id,
                  CASE WHEN "userAId" = ? THEN "userAAction" ELSE "userBAction" END AS my_action,
                  CASE WHEN "userAId" = ? THEN "userBAction" ELSE "userAAction" END AS their_action,
                  COALESCE("isMutual", false) AS is_mutual
                FROM "Match"
                WHERE COALESCE("eventId", '') = ''
                  AND (? = "userAId" OR ? = "userBId")
                """;

        Set<String> actioned = new HashSet<>();
        Set<String> incomingLikes = new HashSet<>();
        Set<String> mutual = new HashSet<>();

        jdbcTemplate.query(
                sql,
                rs -> {
                    String otherId = rs.getString("other_id");
                    String myAction = rs.getString("my_action");
                    String theirAction = rs.getString("their_action");
                    boolean isMutual = rs.getBoolean("is_mutual");

                    if (otherId == null || otherId.isBlank()) {
                        return;
                    }

                    if (isMutual) {
                        mutual.add(otherId);
                    }
                    if (myAction != null && !myAction.isBlank()) {
                        actioned.add(otherId);
                    }
                    if (!isMutual
                            && (myAction == null || myAction.isBlank())
                            && isLikeAction(theirAction)) {
                        incomingLikes.add(otherId);
                    }
                },
                userId,
                userId,
                userId,
                userId,
                userId
        );

        return new GlobalMatchContext(actioned, incomingLikes, mutual);
    }

    private static boolean isLikeAction(String action) {
        if (action == null) {
            return false;
        }
        String normalized = action.trim().toUpperCase(Locale.ROOT);
        return "LIKE".equals(normalized) || "SUPER_LIKE".equals(normalized);
    }

    public List<UserDto> findMatches(
            String userId,
            double userLat,
            double userLon,
            double radiusKm,
            int limit,
            Integer minAge,
            Integer maxAge
    ) {
        double earthRadiusKm = 6371.0;
        double latChange = (radiusKm / earthRadiusKm) * (180.0 / Math.PI);
        double lonChange = (radiusKm / (earthRadiusKm * Math.cos((userLat * Math.PI) / 180.0))) * (180.0 / Math.PI);

        double minLat = userLat - latChange;
        double maxLat = userLat + latChange;
        double minLon = userLon - lonChange;
        double maxLon = userLon + lonChange;

        StringBuilder sql = new StringBuilder(
            "SELECT u.* FROM users.\"User\" u WHERE u.id <> ? " +
                "AND u.\"isOnboardingCompleted\" = true " +
                "AND u.\"isProfileVisible\" = true " +
                "AND u.\"showInMatches\" = true " +
                "AND (" +
                "    u.\"incognitoMode\" = false " +
                "    OR EXISTS (" +
                "        SELECT 1 FROM \"Match\" m " +
                "        WHERE (m.\"userAId\" = u.id AND m.\"userBId\" = ? AND m.\"userAAction\" = 'LIKE') " +
                "           OR (m.\"userBId\" = u.id AND m.\"userAId\" = ? AND m.\"userBAction\" = 'LIKE')" +
                "    )" +
                ") " +
                "AND u.\"lastLatitude\" BETWEEN ? AND ? " +
                "AND u.\"lastLongitude\" BETWEEN ? AND ? " +
                "AND NOT EXISTS (" +
                "SELECT 1 FROM users.\"UserBlock\" b WHERE (b.\"blockerId\" = ? AND b.\"targetUserId\" = u.id) OR (b.\"blockerId\" = u.id AND b.\"targetUserId\" = ?)) " +
                "AND NOT EXISTS (" +
                "SELECT 1 FROM \"Match\" m " +
                "WHERE COALESCE(m.\"eventId\", '') = '' " +
                "AND (" +
                "  (m.\"userAId\" = ? AND m.\"userBId\" = u.id AND m.\"userAAction\" IS NOT NULL) " +
                "  OR (m.\"userBId\" = ? AND m.\"userAId\" = u.id AND m.\"userBAction\" IS NOT NULL) " +
                "  OR (COALESCE(m.\"isMutual\", false) = true AND ((m.\"userAId\" = ? AND m.\"userBId\" = u.id) OR (m.\"userBId\" = ? AND m.\"userAId\" = u.id)))" +
                ")) "
        );

        List<Object> params = new ArrayList<>();
        params.add(userId);
        params.add(userId);
        params.add(userId);
        params.add(minLat);
        params.add(maxLat);
        params.add(minLon);
        params.add(maxLon);
        params.add(userId);
        params.add(userId);
        params.add(userId);
        params.add(userId);
        params.add(userId);
        params.add(userId);

        if (minAge != null) {
            sql.append("AND age >= ? ");
            params.add(minAge);
        }
        if (maxAge != null) {
            sql.append("AND age <= ? ");
            params.add(maxAge);
        }

        sql.append("LIMIT ?");
        params.add(limit);

        return jdbcTemplate.query(sql.toString(), mapper(), params.toArray());
    }

    private RowMapper<UserDto> mapper() {
        return (rs, rowNum) -> mapUser(rs);
    }

    private UserDto mapUser(ResultSet rs) throws SQLException {
        return mapUser(rs, null, null);
    }

    private UserDto mapUser(ResultSet rs, Double averageRating, Long eventsCreatedCount) throws SQLException {
        Array interestsArr = rs.getArray("interests");
        List<String> interests = interestsArr == null ? List.of() : Arrays.asList((String[]) interestsArr.getArray());

        Array photosArr = rs.getArray("photos");
        List<String> photos = photosArr == null ? List.of() : Arrays.asList((String[]) photosArr.getArray());

        Map<String, Object> socialLinks = null;
        String socialLinksJson = rs.getString("socialLinks");
        if (socialLinksJson != null) {
            try {
                socialLinks = objectMapper.readValue(socialLinksJson, new TypeReference<>() {});
            } catch (Exception ignored) {
                socialLinks = null;
            }
        }

        return new UserDto(
                rs.getString("id"),
                rs.getString("firebaseUid"),
                rs.getString("email"),
                rs.getString("displayName"),
                rs.getString("photoUrl"),
                rs.getString("coverImageUrl"),
                photos,
                rs.getString("bio"),
                interests,
                socialLinks,
                (Integer) rs.getObject("age"),
                rs.getString("gender"),
                rs.getString("role"),
                (Double) rs.getObject("lastLatitude"),
                (Double) rs.getObject("lastLongitude"),
                toInstant(rs.getObject("lastLocationUpdate")),
                (Boolean) rs.getObject("isProfileVisible"),
                (Boolean) rs.getObject("isLocationVisible"),
                (Integer) rs.getObject("minAge"),
                (Integer) rs.getObject("maxAge"),
                (Integer) rs.getObject("maxDistance"),
                (Boolean) rs.getObject("showVisitedEvents"),
                (Boolean) rs.getObject("showInMatches"),
                (Boolean) rs.getObject("incognitoMode"),
                (Boolean) rs.getObject("hideOnlineStatus"),
                rs.getString("fcmToken"),
                (Boolean) rs.getObject("isOnboardingCompleted"),
                toInstant(rs.getObject("createdAt")),
                toInstant(rs.getObject("updatedAt")),
                averageRating,
                eventsCreatedCount
        );
    }

    /**
     * Find a user by ID and enrich with organizer rating from events schema.
     * averageRating/eventsCreatedCount are null if user has fewer than 3 approved events.
     */
    public Optional<UserDto> findByIdWithRating(String id) {
        List<UserDto> rows = jdbcTemplate.query(
                "SELECT u.*, " +
                "  COUNT(DISTINCT e.id) AS events_count, " +
                "  CASE WHEN COUNT(DISTINCT e.id) >= 3 THEN AVG(r.rating) ELSE NULL END AS avg_rating " +
                "FROM users.\"User\" u " +
                "LEFT JOIN events.\"Event\" e ON e.\"createdById\" = u.id AND e.status = 'APPROVED' " +
                "LEFT JOIN events.\"EventRating\" r ON r.\"eventId\" = e.id " +
                "WHERE u.id = ? " +
                "GROUP BY u.id",
                (rs, rn) -> {
                    long eventsCount = rs.getLong("events_count");
                    Object val = rs.getObject("avg_rating");
                    Double avgRating = null;
                    if (eventsCount >= 3 && val != null) {
                        if (val instanceof java.math.BigDecimal bd) {
                            avgRating = bd.doubleValue();
                        } else if (val instanceof Double d) {
                            avgRating = d;
                        } else {
                            avgRating = rs.getDouble("avg_rating");
                        }
                    }
                    Long eventsCreatedCount = eventsCount > 0 ? eventsCount : null;
                    try {
                        return mapUser(rs, avgRating, eventsCreatedCount);
                    } catch (Exception e) {
                        throw new RuntimeException(e);
                    }
                },
                id
        );
        return rows.stream().findFirst();
    }

    public void addPhoto(String userId, String photoUrl) {
        jdbcTemplate.update(
                "UPDATE users.\"User\" SET \"photos\" = array_append(COALESCE(\"photos\", ARRAY[]::text[]), ?), \"updatedAt\" = NOW() WHERE id = ?",
                photoUrl,
                userId
        );
    }

    public void removePhoto(String userId, String photoUrl) {
        jdbcTemplate.update(
                "UPDATE users.\"User\" SET \"photos\" = array_remove(\"photos\", ?), \"updatedAt\" = NOW() WHERE id = ?",
                photoUrl,
                userId
        );
    }

    public void blockUser(String blockerId, String targetUserId) {
        jdbcTemplate.update(
                "INSERT INTO users.\"UserBlock\" (id, \"blockerId\", \"targetUserId\", \"createdAt\") VALUES (?, ?, ?, NOW()) ON CONFLICT (\"blockerId\", \"targetUserId\") DO NOTHING",
                UUID.randomUUID().toString(),
                blockerId,
                targetUserId
        );
    }

    public void unblockUser(String blockerId, String targetUserId) {
        jdbcTemplate.update(
                "DELETE FROM users.\"UserBlock\" WHERE \"blockerId\" = ? AND \"targetUserId\" = ?",
                blockerId,
                targetUserId
        );
    }

    public UserDto updateRole(String userId, String role) {
        int rowsUpdated = jdbcTemplate.update(
                "UPDATE users.\"User\" SET \"role\" = CAST(? AS users.\"UserRole\"), \"updatedAt\" = NOW() WHERE id = ?",
                role,
                userId
        );

        if (rowsUpdated == 0) {
            throw new UserService.NotFoundException("User not found");
        }

        return findById(userId).orElseThrow();
    }

    private Instant toInstant(Object ts) {
        if (ts == null) return null;
        if (ts instanceof java.sql.Timestamp t) return t.toInstant();
        if (ts instanceof java.time.OffsetDateTime odt) return odt.toInstant();
        if (ts instanceof java.time.LocalDateTime ldt) return ldt.atZone(java.time.ZoneOffset.UTC).toInstant();
        return null;
    }
}
